lib/inspec/config.rb

```diff
@@ -0,0 +1,286 @@
+      options.merge!(config_file_cli_options)
+      # Reporter options may be defined top-level.
+      options.merge!(config_file_reporter_options)
+
+      # Highest precedence: merge in any options defined via the CLI
+      options.merge!(@cli_opts)
+
+      options
+    end
+
+    #-----------------------------------------------------------------------#
+    #                          Finalization
+    #-----------------------------------------------------------------------#
+    def finalize_options
+      options = @merged_options.dup
+
+      finalize_set_top_level_command(options)
+      finalize_parse_reporters(options)
+      finalize_handle_sudo(options)
+      finalize_compliance_login(options)
+
+      Thor::CoreExt::HashWithIndifferentAccess.new(options)
+    end
+
+    def finalize_set_top_level_command(options)
+      options[:type] = @command_name
+      Inspec::BaseCLI.inspec_cli_command = @command_name # TODO: move to a more relevant location
+    end
+
+    def finalize_parse_reporters(options) # rubocop:disable Metrics/AbcSize
+      # default to cli report for ad-hoc runners
+      options['reporter'] = ['cli'] if options['reporter'].nil?
+
+      # parse out cli to proper report format
+      if options['reporter'].is_a?(Array)
+        reports = {}
+        options['reporter'].each do |report|
+          reporter_name, destination = report.split(':', 2)
+          if destination.nil? || destination.strip == '-'
+            reports[reporter_name] = { 'stdout' => true }
+          else
+            reports[reporter_name] = {
+              'file' => destination,
+              'stdout' => false,
+            }
+            reports[reporter_name]['target_id'] = options['target_id'] if options['target_id']
+          end
+        end
+        options['reporter'] = reports
+      end
+
+      # add in stdout if not specified
+      if options['reporter'].is_a?(Hash)
+        options['reporter'].each do |reporter_name, config|
+          options['reporter'][reporter_name] = {} if config.nil?
+          options['reporter'][reporter_name]['stdout'] = true if options['reporter'][reporter_name].empty?
+          options['reporter'][reporter_name]['target_id'] = options['target_id'] if options['target_id']
+        end
+      end
+
+      validate_reporters(options['reporter'])
+      options
+    end
+
+    def finalize_handle_sudo(options)
+      # Due to limitations in Thor it is not possible to set an argument to be
+      # both optional and its value to be mandatory. E.g. the user supplying
+      # the --password argument is optional and not always required, but
+      # whenever it is used, it requires a value. Handle options that were
+      # defined in such a way and require a value here:
+      %w{password sudo-password}.each do |option_name|
+        snake_case_option_name = option_name.tr('-', '_').to_s
+        next unless options[snake_case_option_name] == -1 # Dubious - does thor always set -1 for missing value?
```

Comment:
This one goes a bit back, context can be found here:

https://github.com/inspec/inspec/pull/1918

https://github.com/inspec/inspec/pull/1918/files#diff-2b20044866abd29a7a781f5760238403R30

I think we should link it or summarize it here.